src/device-state/db-format.ts

```diff
@@ -0,0 +1,185 @@
+	const dbApp = await getDBEntry(id);
+	return await buildApp(dbApp);
+}
+
+export async function getApps(): Promise<InstancedAppState> {
+	const dbApps = await getDBEntry();
+	const apps: InstancedAppState = {};
+	for (const app of dbApps) {
+		apps[app.appId] = await buildApp(app);
+	}
+	return apps;
+}
+
+async function buildApp(dbApp: targetStateCache.DatabaseApp) {
+	const volumes = _.mapValues(JSON.parse(dbApp.volumes) ?? {}, (conf, name) => {
+		if (conf == null) {
+			conf = {};
+		}
+		if (conf.labels == null) {
+			conf.labels = {};
+		}
+		return Volume.fromComposeObject(name, dbApp.appId, conf);
+	});
+
+	const networks = _.mapValues(
+		JSON.parse(dbApp.networks) ?? {},
+		(conf, name) => {
+			if (conf == null) {
+				conf = {};
+			}
+			return Network.fromComposeObject(name, dbApp.appId, conf);
+		},
+	);
+
+	const opts = await config.get('extendedEnvOptions');
+	const supervisorApiHost = dockerUtils
+		.getNetworkGateway(constants.supervisorNetworkInterface)
+		.catch(() => '127.0.0.1');
+	const hostPathExists = {
+		firmware: await pathExistsOnHost('/lib/firmware'),
+		modules: await pathExistsOnHost('/lib/modules'),
+	};
+	const hostnameOnHost = _.trim(
+		await fs.readFile(
+			path.join(constants.rootMountPoint, '/etc/hostname'),
+			'utf8',
+		),
+	);
+
+	const svcOpts = {
+		appName: dbApp.name,
+		supervisorApiHost,
+		hostPathExists,
+		hostnameOnHost,
+		...opts,
+	};
+
+	// In the db, the services are an array, but here we switch them to an
+	// object so that they are consistent
+	const services = _.keyBy(
+		await Promise.all(
+			(JSON.parse(dbApp.services) ?? []).map(
+				async (svc: ServiceComposeConfig) => {
+					// Try to fill the image id if the image is downloaded
+					let imageInfo: ImageInspectInfo | undefined;
+					try {
+						imageInfo = await images.inspectByName(svc.image);
+					} catch (e) {
+						if (NotFoundError(e)) {
+							imageInfo = undefined;
+						} else {
+							throw e;
+						}
+					}
+
+					const thisSvcOpts = {
+						...svcOpts,
+						imageInfo,
+						serviceName: svc.serviceName,
+					};
+					// We force the casting here as we know that the UUID exists, but the typings do
+					// not
+					return Service.fromComposeObject(
+						svc,
+						(thisSvcOpts as unknown) as DeviceMetadata,
+					);
+				},
+			),
+		),
+		'serviceId',
+	) as Dictionary<Service>;
+
+	return {
+		appId: dbApp.appId,
+		commit: dbApp.commit,
+		releaseId: dbApp.releaseId,
+		name: dbApp.name,
+		source: dbApp.source,
+
+		services,
+		volumes,
+		networks,
+	};
+}
+
+export async function setApps(
+	apps: { [appId: number]: TargetApplication },
+	source: string,
+	trx?: db.Transaction,
+) {
+	const cloned = _.cloneDeep(apps);
+
+	const dbApps = await Promise.all(
+		Object.keys(cloned).map(async (appIdStr) => {
+			const appId = checkInt(appIdStr)!;
+
+			const app = cloned[appId];
+			const services = await Promise.all(
+				_.map(app.services, async (s, sId) => ({
```

Comment:
I know this is an pre-existing issue, but it does mean that services in the db format are an array whereas in the api version they're an object, which is fairly confusing and makes the types awkward as some places expect one style and other expect the other

Reply:
Yeah, I was thinking about making that change to the database, but I'd rather do this first.